var/spack/repos/builtin/packages/hdf5-vol-async/package.py

```diff
@@ -27,10 +27,7 @@ class Hdf5VolAsync(CMakePackage):
     depends_on("hdf5@1.13: +mpi +threadsafe")
 
     def setup_run_environment(self, env):
-        env.set("HDF5_PLUGIN_PATH", self.spec.prefix.lib)
-        vol_connector = "async under_vol=0;under_info=[]"
-        env.set("HDF5_VOL_CONNECTOR", vol_connector)
-        env.set("MPICH_MAX_THREAD_SAFETY", "multiple")
+        env.prepend_path("HDF5_PLUGIN_PATH", self.spec.prefix.lib)
```

Comment:
@lrknox for each of these packages we're changing `env.set` to `env.prepend_path` for `HDF5_PLUGIN_PATH`.  Otherwise, whichever VOL package is loaded last will overwrite the variable (meaning only one VOL would ever be in this variable).

What is not clear is the role of `HDF5_VOL_CONNECTOR`, in theory that could be any value?  This is a slight point of tension, but the motivation for removing this (and `MPICH_MAX_THREAD_SAFETY`) is similar to `set` vs `prepend_path` -- my belief is that the connector needs to be controlled at the application level, this package setting this value makes it so that no other VOL would be used?

Reply:
Not any value, but it could be changed to combine multiple VOL connectors. We can safely assume that if a user loads a connector, they want to use it without much hassle, so those should be set. If they want to override the default behavior by combining multiple connectors, they can still do it.

Reply:
Thanks @jeanbez!  I think the confusion is that we only have one example of `HDF5_VOL_CONNECTOR`.  Consider if both `hdf5-vol-async` and `hdf5-vol-cache` both needed to set that variable.

> but it could be changed to combine multiple VOL connectors.

So it would look like `HDF5_VOL_CONNECTOR=async under_vol=0;under_info=[]; additional_magic=11; etc=[]`.  It's not clear how to approach this problem correctly, other than that `env.set` is a problem.  For example, if I am a user and I have `HDF5_VOL_CONNECTOR=<my custom value>`, the current spack recipe just obliterates it and chooses its own value.  How to set that correctly in `spack` is not currently well defined (but we could possibly try and e.g., add it in if it already exists?  hard part is removing from environment).

Does the problem make more sense now?  The adios2 vol is the only other one not on this PR, it just prepends to `HDF5_PLUGIN_PATH` but daos is also planned (if you have any knowledge of these / they factor in here).

From my perspective, this package should not be allowed to force the VOL connector on people -- we want to make all the VOLs _available_ in the SDK.  If by installing one which sets `HDF5_VOL_CONNECTOR` that implicitly means it's always used, that isn't OK.  Is that what happens?  If it's not clear, I don't have a good understanding of the overall layout here with this variable so we're trying to figure it out :slightly_smiling_face:

> Also, other packages already rely on and assume this behavior to hold. So changing this might cause things to break.

Any interested parties we can @?  Silent breakages are never fun, but if the current behavior is incorrect (I believe this is the case), we should find the right way to fix them together.

Putting aside whether or not this change goes through, assume we remove it: what happens to the broken package?  Will it be a runtime error?  Or will it just silently not do something it was expected to do?

Reply:
One thing that may be useful for HDF5 Vol connectors to provide is a script that can be sources to configure the "default" vol connector. This way, users who which to use hdf5-vol-async could load the VOL Async module, and then source a `hdf5-vol-async-default.env`. Likewise for the other vol connectors.

Here is an example that does some possibly useful checks, there may also be more useful details or configurations that specific vol connects would want.

```sh
# hdf5-vol-async-default.env
echo "Loading HDF5 VOL Async..."

# Save current vol connector?
if [[ ! -z $HDF5_VOL_CONNECTOR ]]; then
  echo "Overriding current VOL Connector config: $HDF5_VOL_CONNECTOR"
  echo "Saving in HDF5_VOL_CONNECT_BAK"
  export HDF5_VOL_CONNECTOR_BAK=$HDF5_VOL_CONNECTOR
fi


# Validate that the target vol connector is in the path
IFS=:
for path in $HDF5_PLUGIN_PATH; do
  # Search for VOL Async plugin, if it isn't in the path then report an error to the user?
done

# Setup the default config
export HDF5_VOL_CONNECTOR="async under_vol=0;under_info=[]"
```

In theory, HDF5 could also provide a VOL preamble script that handles backing up `HDF5_VOL_CONNECTOR` and searching for the expected plugin library in the `HDF5_PLUGIN_PATH`.

Reply:
Marking as resolved per https://github.com/spack/spack/pull/33393#issuecomment-1291051774, the spack package will make all VOLs added here (async, cache, log) _available_ via `HDF5_PLUGIN_PATH` and we will remove `HDF5_VOL_CONNECTOR` getting set since this changes the application behavior without user consent.